_core/smart/functions.md

```diff
@@ -0,0 +1,151 @@
+
+## List operations and functions
+
+Lists may be multi-dimensional. However, note that runtime admission checks on typed function-parameters and data-map functions like `set-entry!` are charged based on the _maximal_ size of the multi-dimensional list.
+
+You can call `filter` `map` and `fold` functions with user-defined functions (that is, functions defined with `(define ...)`, `(define-read-only ...)`, or `(define-public ...)`) or simple, native functions (for example, `+`, `-`, `not`).
+
+## Intra-contract calls
+
+A smart contract may call functions from other smart contracts using a `(contract-call!)` function:
+
+```cl
+(contract-call! contract-name function-name arg0 arg1 ...)
+```
+
+This function accepts a function name and the smart contract's name as input. For example, to call the function `token-transfer` in the smart contract, you would use:
+
+`(contract-call! tokens token-transfer burn-address name-price))`
+
+For intra-contract calls dynamic dispatch is not supported. When a contract is launched, any contracts it depends on (calls) must exist. Additionally, no cycles may exist in the call graph of a smart contract. This prevents recursion (and re-entrancy bugs. A static analysis of the call graph detects such structures and they are rejected by the network.
+
+A smart contract may not modify other smart contracts' data directly; it can read data stored in those smart contracts' maps. This read ability does not alter any confidentiality guarantees of Clarity. All data in a smart contract is inherently public, andis readable through querying the underlying database in any case.
+
+To read another contract's data, use `(fetch-contract-entry)` function. This behaves identically to `(fetch-entry)`, though it accepts a contract principal as an argument in addition to the map name:
+
+```cl
+(fetch-contract-entry
+  'contract-principal
+  'map-name
+  'key-tuple) ;; value tuple or none
+```
+
+For example, you could do this:
+
+```cl
+(fetch-contract-entry
+ 'SC3H92H297DX3YDPFHZGH90G8Z4NPH4VE8E83YWAQ
+ 'name-map
+ 1)     ;;returns owner principal of name represent by integer 12234
+```
+
+Just as with the `(contract-call)` function, the map name and contract principal arguments must be constants, specified at the time of publishing.
+
+Finally, and importantly, the `tx-sender` variable does not change during inter-contract calls. This means that if a transaction invokes a function in a given smart contract, that function is able to make calls into other smart contracts on your behalf. This enables a wide variety of applications, but it comes with some dangers for users of smart contracts. However, the static analysis guarantees of Clarity allow clients to know a priori which functions a given smart contract will ever call. Good clients should always warn users about any potential side effects of a given transaction.
+
+### Reading from Other Smart Contracts
```

Comment:
This subsection looks like a repeat of the above.

Reply:
yep.